internal/providers/github/ghcr/ghcr.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ghcr provides a client for interacting with the GitHub Container Registry
+package ghcr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/google/go-github/v56/github"
+	"golang.org/x/oauth2"
+
+	"github.com/stacklok/minder/internal/verifier/verifyif"
+	provifv1 "github.com/stacklok/minder/pkg/providers/v1"
+)
+
+// GHCR is the struct that contains the GHCR client
+type GHCR struct {
+	client    *github.Client
+	namespace string
+}
+
+// Ensure that the GHCR client implements the ContainerLister interface
+var _ provifv1.ImageLister = (*GHCR)(nil)
+
+// New creates a new GHCR client
+func New(cred provifv1.OAuth2TokenCredential, namespace string) *GHCR {
+	tc := oauth2.NewClient(context.Background(), cred.GetAsOAuth2TokenSource())
+
+	return &GHCR{
+		client:    github.NewClient(tc),
+		namespace: namespace,
+	}
+}
+
+// FromGitHubClient creates a new GHCR client from an existing GitHub client
+func FromGitHubClient(client *github.Client, namespace string) *GHCR {
+	return &GHCR{
+		client:    client,
+		namespace: namespace,
+	}
+}
+
+// ListImages lists the containers in the GHCR
+func (g *GHCR) ListImages(ctx context.Context) ([]string, error) {
+	pageNumber := 0
+	itemsPerPage := 100
+	pt := string(verifyif.ArtifactTypeContainer)
+	opt := &github.PackageListOptions{
```

Comment:
Ah, this is because we want to check `/v2/_catalog`, which we don't have the correct token type for?

https://toddysm.com/2024/02/12/authenticating-with-oci-registries-github-container-registry-ghcr-implementation/ has some hints, and suggests that you can use a GitHub auth token (As `Authorization: Bearer ...`) with the correct scopes to get a GHCR-specific token from `/v2/token` and then use the token from `/v2/token` to call `/v2/_catalog` to get a list of repositories / images.

Note that the blog post suggests that this includes a list of _all_ public repositories in the `/v2/_catalog` endpoint when called when human credentials.  I don't know what the options for GitHub App credentials here are.

Reply:
With the knowledge we have now of how the GitHub App works, we should probably revisit this implementation. I'd like GHCR to work in the following scenarios:
* As part of the GitHub App provider
* As part of the OAuth2 provider
* As a standalone provider